snap/quota/quota.go

```diff
@@ -259,39 +259,34 @@ func ResolveCrossReferences(grps map[string]*Group) error {
 
 // tree recursively returns all of the sub-groups of the group and the group
 // itself.
-func (grp *Group) tree(acc []*Group) ([]*Group, error) {
-	// make sure that none of the sub-groups seen already are present in the
-	// sub-groups of this group
+func (grp *Group) visitTree(visited map[*Group]bool) error {
+	// be paranoid about cycles here and check that none of the sub-groups here
+	// has already been seen before recursing
 	for _, sub := range grp.subGroups {
-		// check if we have already seen this sub-group
-		for _, alreadySeen := range acc {
-			if sub == alreadySeen {
-				return nil, fmt.Errorf("internal error: circular reference found")
-			}
-		}
-
 		// check if this sub-group is actually the same group
 		if sub == grp {
-			return nil, fmt.Errorf("internal error: circular reference found")
+			return fmt.Errorf("internal error: circular reference found")
+		}
+
+		// check if we have already seen this sub-group
+		if _, ok := visited[sub]; ok {
```

Comment:
this can be just visited[sub]  given it's a map to bools?

Reply:
done